packages/llamaindex/src/agent/contextAwareMixin.ts

```diff
@@ -0,0 +1,80 @@
+import { type AgentRunner } from "@llamaindex/core/agent";
+import type {
+  NonStreamingChatEngineParams,
+  StreamingChatEngineParams,
+} from "@llamaindex/core/chat-engine";
+import type { ChatMessage, LLM, MessageContent } from "@llamaindex/core/llms";
+import type { BaseRetriever } from "@llamaindex/core/retriever";
+import type { NodeWithScore } from "@llamaindex/core/schema";
+import { EngineResponse, MetadataMode } from "@llamaindex/core/schema";
+
+type Constructor<T = {}> = new (...args: any[]) => T;
+
+export interface ContextAwareConfig {
+  contextRetriever: BaseRetriever;
+}
+
+export interface ContextAwareAgentRunner extends AgentRunner<LLM> {
+  contextRetriever: BaseRetriever;
+  retrievedContext: string | null;
+  retrieveContext(query: MessageContent): Promise<string>;
+  injectContext(context: string): Promise<void>;
+}
+
+/**
+ * ContextAwareAgentRunner enhances the base AgentRunner with the ability to retrieve and inject relevant context
+ * for each query. This allows the agent to access and utilize appropriate information from a given index or retriever,
+ * providing more informed and context-specific responses to user queries.
+ */
+export function withContextAwareness<T extends Constructor<AgentRunner<LLM>>>(
+  BaseClass: T,
+) {
+  return class extends BaseClass implements ContextAwareAgentRunner {
+    contextRetriever: BaseRetriever;
+    retrievedContext: string | null = null;
+
+    constructor(...args: any[]) {
```

Comment:
I don't recommend any here, it might lose type hint in user side

Reply:
should be `T extends typeof AgentRunner<LLM>, Args extends ConstructorParameter<T>`

and `constructor(...args: Args)`